File: android-native/polaris-study-kit/src/main/java/com/polaris/studykit/Citations.kt
package com.polaris.studykit

/**
 * Deterministic citation formatter — the algorithm behind the "Citation" feature.
 * Applies each style's fixed rule table (APA / MLA / Chicago) to the source fields.
 */
data class Source(
    val authors: List<String> = emptyList(),   // "Last, First" each
    val title: String = "",
    val year: String = "",
    val container: String = "",                 // journal / book / site
    val publisher: String = "",
    val url: String = "",
)

enum class CitationStyle { APA, MLA, CHICAGO }

object Citations {

    fun reference(s: Source, style: CitationStyle): String = when (style) {
        CitationStyle.APA -> apa(s)
        CitationStyle.MLA -> mla(s)
        CitationStyle.CHICAGO -> chicago(s)
    }

    fun inText(s: Source, style: CitationStyle): String {
        val name = firstAuthorLast(s) ?: s.title.split(" ").take(2).joinToString(" ")
        return when (style) {
            CitationStyle.APA -> "($name, ${s.year})"
            CitationStyle.MLA -> "($name)"
            CitationStyle.CHICAGO -> "($name ${s.year})"
        }
    }

    private fun apa(s: Source): String {
        val a = joinAuthors(s.authors, ", ", ", & ", initialsOnly = true)
        val sb = StringBuilder()
        if (a.isNotEmpty()) sb.append("$a ")
        sb.append("(${s.year}). ${s.title}.")
        if (s.container.isNotEmpty()) sb.append(" *${s.container}*.")
        if (s.publisher.isNotEmpty()) sb.append(" ${s.publisher}.")
        if (s.url.isNotEmpty()) sb.append(" ${s.url}")
        return collapse(sb.toString())
    }

    private fun mla(s: Source): String {
        val a = joinAuthors(s.authors, ", ", ", and ", initialsOnly = false)
        val parts = mutableListOf<String>()
        if (a.isNotEmpty()) parts.add("$a.")
        parts.add("\"${s.title}.\"")
        val tail = listOfNotNull(
            s.container.ifEmpty { null }?.let { "*$it*" },
            s.publisher.ifEmpty { null }, s.year.ifEmpty { null }, s.url.ifEmpty { null },
        )
        parts.add(tail.joinToString(", ") + ".")
        return collapse(parts.joinToString(" "))
    }

    private fun chicago(s: Source): String {
        val a = joinAuthors(s.authors, ", ", ", and ", initialsOnly = false)
        val sb = StringBuilder()
        if (a.isNotEmpty()) sb.append("$a. ")
        sb.append("${s.title}.")
        if (s.container.isNotEmpty()) sb.append(" *${s.container}*.")
        val pubYear = listOfNotNull(s.publisher.ifEmpty { null }, s.year.ifEmpty { null })
        if (pubYear.isNotEmpty()) sb.append(" ${pubYear.joinToString(", ")}.")
        if (s.url.isNotEmpty()) sb.append(" ${s.url}.")
        return collapse(sb.toString())
    }

    private fun firstAuthorLast(s: Source): String? =
        s.authors.firstOrNull()?.substringBefore(",")?.trim()?.ifEmpty { null }

    private fun joinAuthors(
        authors: List<String>, sep: String, last: String, initialsOnly: Boolean,
    ): String {
        val f = authors.map { if (initialsOnly) initialsForm(it) else it }
        return when (f.size) {
            0 -> ""
            1 -> f[0]
            else -> f.dropLast(1).joinToString(sep) + last + f.last()
        }
    }

    /** "Smith, John Paul" -> "Smith, J. P." */
    private fun initialsForm(name: String): String {
        val parts = name.split(",", limit = 2).map { it.trim() }
        if (parts.size != 2) return name
        val initials = parts[1].split(" ").filter { it.isNotEmpty() }
            .joinToString(" ") { "${it.first()}." }
        return "${parts[0]}, $initials"
    }

    private fun collapse(s: String): String =
        s.replace("  ", " ").replace(" .", ".").trim()
}

File: ios-native/Sources/PolarisStudyKit/Citations.swift
import Foundation

/// Deterministic citation formatter — the algorithm behind the "Citation" feature.
///
/// No AI: builds APA / MLA / Chicago references by applying each style's fixed rule table to
/// the source's fields (author, title, year, …). Also produces the matching in-text citation.
public struct Source: Sendable {
    public var authors: [String]   // "Last, First" each
    public var title: String
    public var year: String
    public var container: String    // journal / book / site name
    public var publisher: String
    public var url: String

    public init(authors: [String] = [], title: String = "", year: String = "",
                container: String = "", publisher: String = "", url: String = "") {
        self.authors = authors
        self.title = title
        self.year = year
        self.container = container
        self.publisher = publisher
        self.url = url
    }
}

public enum CitationStyle: String, CaseIterable, Sendable { case apa, mla, chicago }

public enum Citations {

    public static func reference(_ s: Source, style: CitationStyle) -> String {
        switch style {
        case .apa: return apa(s)
        case .mla: return mla(s)
        case .chicago: return chicago(s)
        }
    }

    /// In-text citation, e.g. "(Smith, 2020)" / "(Smith)" / "(Smith 2020)".
    public static func inText(_ s: Source, style: CitationStyle) -> String {
        let name = firstAuthorLast(s) ?? shortTitle(s)
        switch style {
        case .apa: return "(\(name), \(s.year))"
        case .mla: return "(\(name))"
        case .chicago: return "(\(name) \(s.year))"
        }
    }

    // MARK: - Styles

    private static func apa(_ s: Source) -> String {
        // Authors (Last, F. M.) & final "&"; (Year). Title. Container. Publisher. URL
        let a = joinAuthors(s.authors, sep: ", ", last: ", & ", initialsOnly: true)
        var out = a.isEmpty ? "" : "\(a) "
        out += "(\(s.year)). \(s.title)."
        if !s.container.isEmpty { out += " \(italicHint(s.container))." }
        if !s.publisher.isEmpty { out += " \(s.publisher)." }
        if !s.url.isEmpty { out += " \(s.url)" }
        return collapse(out)
    }

    private static func mla(_ s: Source) -> String {
        // Author. "Title." Container, Publisher, Year, URL.
        let a = joinAuthors(s.authors, sep: ", ", last: ", and ", initialsOnly: false)
        var parts: [String] = []
        if !a.isEmpty { parts.append("\(a).") }
        parts.append("\"\(s.title).\"")
        var tail: [String] = []
        if !s.container.isEmpty { tail.append(italicHint(s.container)) }
        if !s.publisher.isEmpty { tail.append(s.publisher) }
        if !s.year.isEmpty { tail.append(s.year) }
        if !s.url.isEmpty { tail.append(s.url) }
        parts.append(tail.joined(separator: ", ") + ".")
        return collapse(parts.joined(separator: " "))
    }

    private static func chicago(_ s: Source) -> String {
        // Author. Title. Container. Publisher, Year. URL.
        let a = joinAuthors(s.authors, sep: ", ", last: ", and ", initialsOnly: false)
        var out = a.isEmpty ? "" : "\(a). "
        out += "\(s.title)."
        if !s.container.isEmpty { out += " \(italicHint(s.container))." }
        var pubYear: [String] = []
        if !s.publisher.isEmpty { pubYear.append(s.publisher) }
        if !s.year.isEmpty { pubYear.append(s.year) }
        if !pubYear.isEmpty { out += " \(pubYear.joined(separator: ", "))." }
        if !s.url.isEmpty { out += " \(s.url)." }
        return collapse(out)
    }

    // MARK: - Helpers

    private static func firstAuthorLast(_ s: Source) -> String? {
        guard let first = s.authors.first else { return nil }
        return first.split(separator: ",").first.map(String.init)?
            .trimmingCharacters(in: .whitespaces)
    }

    private static func shortTitle(_ s: Source) -> String {
        s.title.split(separator: " ").prefix(2).joined(separator: " ")
    }

    private static func italicHint(_ text: String) -> String { "*\(text)*" }  // markdown italics

    private static func joinAuthors(_ authors: [String], sep: String, last: String,
                                    initialsOnly: Bool) -> String {
        let formatted = authors.map { initialsOnly ? initialsForm($0) : $0 }
        switch formatted.count {
        case 0: return ""
        case 1: return formatted[0]
        default:
            return formatted.dropLast().joined(separator: sep) + last + formatted.last!
        }
    }

    /// "Smith, John Paul" -> "Smith, J. P." (APA-style initials).
    private static func initialsForm(_ name: String) -> String {
        let parts = name.split(separator: ",", maxSplits: 1).map {
            $0.trimmingCharacters(in: .whitespaces)
        }
        guard parts.count == 2 else { return name }
        let initials = parts[1].split(separator: " ")
            .compactMap { $0.first.map { "\($0)." } }
            .joined(separator: " ")
        return "\(parts[0]), \(initials)"
    }

    private static func collapse(_ s: String) -> String {
        s.replacingOccurrences(of: "  ", with: " ")
            .replacingOccurrences(of: " .", with: ".")
            .trimmingCharacters(in: .whitespaces)
    }
}
